mate-javascript.md

```diff
@@ -0,0 +1,852 @@
+    - [10.8 Don't shorten variables or functions names](#108-dont-shorten-variables-or-functions-names)
+    - [10.9 If the variable/property/method is a `boolean`, use `isVal()` or `hasVal()`.](#109if-the-variablepropertymethod-is-aboolean-useisvalorhasval)
+- [11. Standard Library](#11-standard-library)
+    - [11.1. Use `Number.isNaN` instead of global `isNaN`.](#111usenumberisnaninstead-of-globalisnan)
+    - [11.2. Use `Number.isFinite` instead of global `isFinite`.](#112usenumberisfiniteinstead-of-globalisfinite)
+- [12. Testing](#12-testing)
+    - [12.1. **Yup.**](#121yup)
+    - [12.2. **No, but seriously**:](#122no-but-seriously)
+
+
+
+1\. Objects
+-----------
+
+
+#### 1.1. Prefer the object spread operator over [`Object.assign`](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/Object/assign) to shallow-copy objects. Use the object rest operator to get a new object with certain properties omitted.
+
+
+
+```javascript
+// very bad
+const original = {
+  a: 1,
+  b: 2
+};
+
+const copy = Object.assign(original, { c: 3 }); // this mutates `original` ಠ_ಠ
+delete copy.a; // so does this
+
+// ❌ bad
+const original = {
+  a: 1,
+  b: 2
+};
+const copy = Object.assign({}, original, { c: 3 });
+// copy => { a: 1, b: 2, c: 3 }
+
+// ✅ good
+const original = { a: 1, b: 2 };
+const copy = { ...original, c: 3 };
+// copy => { a: 1, b: 2, c: 3 }
+
+const { a, ...noA } = copy;
+// noA => { b: 2, c: 3 }
+```
+
+
+2\. Destructuring
+-----------------
+
+#### 2.1. Use object destructuring for multiple return values, not array destructuring.
+
+
+>❓Why? You can add new properties over time or change the order of things without breaking call sites.
+
+```javascript
+// ❌ bad
+function processInput(input) {
+  // then a miracle occurs
+  return [left, right, top, bottom];
+}
+
+// the caller needs to think about the order of return data
+const [left, __, top] = processInput(input);
+
+// ✅ good
+function processInput(input) {
+  // then a miracle occurs
+  return { left, right, top, bottom };
+}
+
+// the caller selects only the data they need
+const { left, top } = processInput(input);
+```
+
+3\. Functions
+-------------
+
+#### 3.1. Avoid side effects with default parameters.
+
+
+>❓Why? They are confusing to reason about.
+
+
+```javascript
+var b = 1;
+
+// ❌ bad
+function count(a = b++) {
+  console.log(a);
+}
+
+count();  // 1
+count();  // 2
+count(3); // 3
+count();  // 3
+```
+
+
+
+#### 3.2. Always put default parameters last.
+
+```javascript
+// ❌ bad
+function handleThings(opts = {}, name) {
+  // ...
+}
+
+// ✅ good
+function handleThings(name, opts = {}) {
+  // ...
+}
+```
+
+
+
+#### 3.3. If parameters number is greater than 2, put them into `options` parameter
```

Comment:
I disagree here. It's ok to have 4 params sometimes.
The more important thing - we shouldn't have 2 params of the same type in a row. Even if only 2 of them. It's really error-prone, in contrary to having 5 different arguments.

Reply:
Can be covered with ESLint.